### PR TITLE
Enable the guest checkout for digi sub gift redemption

### PR DESCRIPTION
**Pull request by i-hardy**
## What are you doing in this PR?

This removes the requirement for a user to be signed in in order to redeem a gift subscription. Instead a user can provide their name and email address in the redemption form and a corresponding Identity account will be found or created for them in the same way as with the guest checkout.

This has involved changes to the redemption form and underlying Redux state to support and validate user input. It's also meant removing the routes for creating/processing a redemption and the thank-you page- moving from the form to processing to the thank you page is now done on the client side, as with the other checkouts.

[**Trello Card**](https://trello.com/c/R4qDueSb)

## Why are you doing this?

This is primarily to fix a rare bug where users who had Identity accounts that did not have an associated first and last name could appear signed in for the purpose of redeeming a gift subscription, but whose user data was rejected by Zuora due to the missing names.

This also brings the redemption form in line with our other checkouts in not requiring users to first be handed off to Identity to sign in or sign up, ahead of upcoming changes to how account creation works.

## Is this an AB test?
- [ ] Yes
- [x] No

## Accessibility test checklist
 - [x] [Tested with screen reader](https://accessibility.gutools.co.uk/testing/web/screen-readers/)
 - [x] [Navigable with keyboard](https://accessibility.gutools.co.uk/testing/web/keyboard-navigation/)
 - [x] [Colour contrast passed](https://accessibility.gutools.co.uk/testing/web/colour-contrast/)

## Screenshots

(Please excuse the invalid code warning, I had just redeemed that code using VoiceOver)

**Mobile**

Screenshot 2021-11-16 at 13 10 36

**Tablet**

Screenshot 2021-11-16 at 13 11 50

**Desktop**

Screenshot 2021-11-16 at 12-58-11 Support the Guardian Redeem your code

**Desktop - showing form validation**

Screenshot 2021-11-16 at 12-58-49 Support the Guardian Redeem your code

### CODE REVIEW

File: support-frontend/assets/helpers/forms/paymentIntegrations/readerRevenueApis.ts [hunk 1]
```diff
@@ -204,18 +198,20 @@ export type PaymentAuthorisation =
...
-	| {
-			paymentStatus: 'success';
-			subscriptionCreationPending: true;
-	  }
-	| {
-			paymentStatus: 'failure';
-			error: ErrorReason;
-	  };
+export type PaymentResult = {
+	paymentStatus: string;
```
Comment (rupertbates) [inline support-frontend/assets/helpers/forms/paymentIntegrations/readerRevenueApis.ts:202]: Should this be of type `Status`?

File: support-frontend/assets/pages/subscriptions-redemption/components/redemptionForm.tsx [hunk 2]
```diff
@@ -29,142 +28,162 @@ import type {
...
 									label="Insert code"
-									css={css`
-										max-width: 300px;
-									`}
 								/>
-							</div>
+							</FormSection>
+							<FormSection title="Your details">
+								<PersonalDetails
+									firstName={props.firstName}
```
Comment (rupertbates) [inline support-frontend/assets/pages/subscriptions-redemption/components/redemptionForm.tsx:159]: Github is really not helping out here is it?

File: support-frontend/assets/pages/subscriptions-redemption/api.ts [hunk 3]
```diff
@@ -93,33 +103,58 @@ function validateUserCode(userCode: string, dispatch: Dispatch<Action>) {
...
+	}
+	return formFieldErrors.length === 0;
+}
+
+function submitCode(
+	dispatch: Dispatch<Action>,
+	state: RedemptionPageState,
+): void {
+	const userCode = state.page.userCode ?? '';
+	if (validateFormFields(dispatch, state)) {
```
[the remark this answers is not in the document]
Comment (i-hardy) [reply support-frontend/assets/pages/subscriptions-redemption/api.ts:126]: Yes, that's a good idea, I'll break it down a bit more.

File: support-frontend/assets/pages/subscriptions-redemption/components/redemptionForm.tsx (hunk 2 above)
Comment (i-hardy) [reply support-frontend/assets/pages/subscriptions-redemption/components/redemptionForm.tsx:159]: I'm not sure what you're seeing, everything looks normal to me.

Comment (rupertbates) [reply support-frontend/assets/pages/subscriptions-redemption/components/redemptionForm.tsx:159]: Everything is wrapping in a weird way, it's fine though - just need to squint a bit

**Review by rupertbates — APPROVED** [review]
👍

**Pull request merged**